Reorganise the devguide to directories

AA-Turner opened the pull request:
From the discord discussion, this is an attempt to add some structure to Docs/Tests/Triage and clean up the top level.

No content has been edited, except for combing two duplicated "how to become a triager" sections. As I moved content between pages I had to edit the underlines -- I also standardised some `^` underlines and out-of-order underlines to allow easier page moves in the future (e.g. a core-dev folder, buildbots, language design and change).

cc: @ezio-melotti @encukou @CAM-Gerlach 

Preview: https://cpython-devguide--895.org.readthedocs.build/

A

hugovk commented:
Where things are renamed (e.g. [`/docquality/`](https://devguide.python.org/docquality/) -> [`/documenting/docquality/`](https://cpython-devguide--895.org.readthedocs.build/documenting/docquality/)) can we put redirects in, so we don't linkrot the web?

AA-Turner commented:
This would need to be at the Fastly level right?

A

CAM-Gerlach commented:
> This would need to be at the Fastly level right?

For specific individual pages like this (rather than patterns), as mentioned before you could do it via manual or autogenerated (via e.g. my script, the `sphinx-redirects` plugin, etc) HTML redirect template pages with `meta http_equiv="refresh"`, which isn't as efficient as doing it at the web server or caching/CDN layer, but makes it practical to configure and maintain without having to ask @ewdurbin every time they are needed, but ideally, yeah. If the Fastly infra redirect config is hosted in the repo, as IIRC Ee did for the PEPs, that's the best of both worlds.

CAM-Gerlach commented on index.rst:
```diff
@@ -156,27 +156,46 @@ Core developers and contributors alike will find the following guides useful:
+* :doc:`testing/silencewarnings`
 * Fixing issues found by the :doc:`buildbots <buildbots>`
 * :doc:`coverity`
 * Helping out with reviewing `open pull requests`_.
```
Two lines below this (too far for me to suggest/comment on it) there's an entry for "Fixing easy issues", which is already included in the table above, and doesn't really make sense under the heading "Advanced topics"

CAM-Gerlach replied:
I don't think we do, unless it was updated drastically, since right now it closely aligns with the categories (and where it doesn't, it probably should).

The same is even more true of the Topic Index Appendix page (I was just going to remove the page completely in my proposed followup; I'd added an item to my review but deleted it before submitting, to not get too into the weeds of what I wasn't sure was in scope here anyway), as aside from missing some of the items added here, it is nearly a 1:1 duplicate of the index page ToC and the sidebar on every page, and thus serves no meaningful purpose other than increasing maintenance burden and the risk of it getting out of sync.

This still leaves us with the sidebar, the full index on the index page (though that's somewhat redundant and probably can also be set to hidden) and several other more focused page listings on individual pages.

CAM-Gerlach commented on index.rst:
```diff
@@ -300,14 +319,10 @@ Full Table of Contents
    fixingissues
-   tracker
-   triaging
+   triaging/index
```
```suggestion
   issues/index
```

Maybe a little bikesheddy, but since issues (reporting, labels and triaging) is the overarching theme here, the index pages talks about issues in general and there's a sub-page already named triaging, wouldn't it make more sense and be less potentially confusing to call the top-level section `issues` rather than `triaging`?

CAM-Gerlach commented on buildbots.rst:
```diff
@@ -105,7 +105,7 @@ introducing additional failures should generally not be an option.
 ------------------------
 
-Sometimes, while you have run the :doc:`whole test suite <runtests>` before
+Sometimes, while you have run the :doc:`whole test suite <testing/index>` before
```
```suggestion
Sometimes, while you have run the :ref:`whole test suite <runtests>` before
```

Isn't it better to link the `:ref:` rather than the `:doc:`, like you did elsewhere, to avoid things breaking if either this document or the testing one is ever moved or renamed again? At the very least, its worth using an absolute rather than relative path so if and when this document gets moved into a subdir, the link won't break (really only makes sense to use relative paths within the same subdirectory tree).

CAM-Gerlach commented on index.rst:
```diff
@@ -42,7 +42,7 @@ instructions please see the :ref:`setup guide <setup>`.
    :ref:`Mac OS <MacOS>`, and :ref:`Windows <windows-compiling>`.
 
-4. :doc:`Run the tests <runtests>`::
+4. :doc:`Run the tests <testing/index>`::
```
```suggestion
4. :ref:`Run the tests <runtests>`::
```

Same here, no?

CAM-Gerlach commented on pullrequest.rst:
```diff
@@ -89,7 +90,7 @@ You should have already :ref:`set up your system <setup>`,
      ./python -m test
 
-  (Learn more about :ref:`patchcheck` and about :doc:`runtests`)
+  (Learn more about :ref:`patchcheck` and about :doc:`testing/index`)
```
```suggestion
  (Learn more about :ref:`patchcheck` and about :ref:`runtests`)
```

And here

CAM-Gerlach reviewed (commented):
I think this is a good start over what is there now. To note, while it still leaves over_thirty_ (30!) top-level headings in the ToC/sidebar with an not always clear order, it still is a net improvement over the 35 that were there before. After reading through all the existing pages in detail and taking the changes in this PR as a starting point, I've developed a proposed structure that would organize everything under 8 top-level headings, with a logical, thoughtful order at both levels. 

Here's what that looks like, for reference:

<details>

<summary>Proposed re-organization</summary>

Getting Started (existing page)
- Fixing Easy Issues and Beyond
- Git bootcamp and cheat sheet
- Lifecycle of a Pull Request
- Where to Get Help

Development workflow (Maybe summary material pulled from index page?)
- Following Python's development
- Development cycle
- Adding to the stdlib
- Developing stdlib extension modules
- Modifying Python's C API
- Changing the Python language
- Porting Python to a new platform

Issue tracking and triage
- Triaging an issue
- GitHub labels
- GitHub issues for BPO users
- Python triage team

Documenting Python
- Helping with documentation
- Documentation Markup
- Style guide
- Translating
- Helping with the developer's guide

Testing and CI
- Increase test coverage
- Silence warnings in the test suite
- Continuous Integration
- Running a buildbot worker

Core Developers ("What it means" section of How to Become a Core Developer page)
- How to become a core developer (rest of Core Developer page)
- Accepting pull requests
- Experts Index
- Core developer motivations & affiliations
- Developer log

CPython internals (Exploring page)
- Grammar
- Parser
- Compiler
- GC

Advanced tools (brief bullet-point summary)
- gdb support
- Dynamic analysis with Clang
- Coverity scan

</details>

I was initially going to propose making them here to keep things atomic and coherent and minimize noise and churn as much as practicable, but given that would be a major change to an already big PR, the changes here are good on their own and the above proposal doesn't further change anything done here aside from adding the "CI" and "Buildbot worker" docs to the "Testing and CI" section, it probably makes more sense to just merge this as-is (or with just those two added) and I'll propose the rest in a followup where we can discuss it further.

hugovk commented on documenting/styleguide.rst:
```diff
@@ -0,0 +1,220 @@
+
+In the Python documentation, the use of sentence case in section titles is
+preferable, but consistency within a unit is more important than
+following this rule.  If you add a section to a chapter where most
+sections are in title case, you can either convert all titles to
+sentence case or use the dominant style in the new section title.
```
Re: this advice:

At least for the files you're moving around, shall we convert the titles to sentence case?

CAM-Gerlach replied:
If we had the choice now, sentence case would be preferred for consistency with the Python docs, but I'm not sure its worth the churn changing at this point. We should be consistent, though (as this section says), and there are some that use sentence case, and at least one title that mixes both.

hugovk commented on documenting/translating.rst:
```diff
@@ -0,0 +1,254 @@
+  your language to be added in the language picker on docs.python.org.
+
+
+PEP 545 summary:
+================
```
```suggestion
PEP 545 summary
===============
```

ewdurbin commented:
> > This would need to be at the Fastly level right?
> 
> For specific individual pages like this (rather than patterns), as mentioned before you could do it via manual or autogenerated (via e.g. my script, the `sphinx-redirects` plugin, etc) HTML redirect template pages with `meta http_equiv="refresh"`, which isn't as efficient as doing it at the web server or caching/CDN layer, but makes it practical to configure and maintain without having to ask @ewdurbin every time they are needed, but ideally, yeah. If the Fastly infra redirect config is hosted in the repo, as IIRC Ee did for the PEPs, that's the best of both worlds.

devguide.python.org is hosted directly with readthedocs.org, so Fastly is not involved.

RTD supports configuring redirects. Currently Brett Cannon, Carol Willing, Julien Palard, Mariatta, and myself are maintainers on the RTD project, and can configure such redirects per [the documentation](https://docs.readthedocs.io/en/stable/user-defined-redirects.html).

AA-Turner commented:
The syntax seems to be:

```
Type: Page Redirect
From URL: /page
To URL: /new/path/page
```

so if we compile a list of page moves, we should ping one of the admins to add it to the config when this PR is merged.

AA-Turner commented:
Sorry for force-pushing, I made enough change to basically wipe out the entire previous branch history.

Implemented the structure across every page in the devguide, updated all references, split various pages off, and updated language to account for the change in structure.

A

AA-Turner commented:
Rendered version: https://cpython-devguide--895.org.readthedocs.build/

A

AA-Turner commented:
[Cam]
> Minimizes noise, churn and disruption and makes it easier to ignore/skip in Git blame, diffs, logs, since it will be one atomic PR squashed to one commit

I would suggest that this PR is merged normally rather than squash merged -- as I've edited files and split out sections, git may have a hard time tracking file moves, whereas it does a pretty good job on my branch. (I have git blame configured to detect movements within and across files.)

A

CAM-Gerlach commented:
I'll have a full review a bit later; its a lot to go through.

> RTD supports configuring redirects. Currently Brett Cannon, Carol Willing, Julien Palard, Mariatta, and myself are maintainers on the RTD project, and can configure such redirects per [the documentation](https://docs.readthedocs.io/en/stable/user-defined-redirects.html).

Thanks, @ewdurbin . I assume there's no easy way to host this configuration in the repo itself, as you did for Fastly on the PEPs repo?

> I would suggest that this PR is merged normally rather than squash merged -- as I've edited files and split out sections, git may have a hard time tracking file moves, whereas it does a pretty good job on my branch. (I have git blame configured to detect movements within and across files.)

@AA-Turner That would be a question for the devguide maintainers, but at least in GitHub's Files pane, all the true or close moves are detected correctly, while splits are considered new files (which, in effect, they are). So it _looks_ like things are more or less correct; you could do a test locally to confirm.

CAM-Gerlach commented on testing/index.rst:
```diff
@@ -0,0 +1,12 @@
+====================================
+ Testing and Continuous Integration
+====================================
```
```suggestion
===================
Testing & Buildbots
===================
```

I was tricked by the original title of the CI page; this is actually specifically about buildbots, not CI in general (or specifically on cloud CI providers, as most OSS developers would assume when reading this). This change another benefit—it makes the title short enough to fit on one line, which means all the category titles do so (which is really nice).

Additionally, use `&` which is shorter and consistent with what is used in the page titles.

CAM-Gerlach commented on testing/runtests.rst:
```diff
@@ -1,7 +1,8 @@
 
-Running & Writing Tests
-=======================
+=========================
+ Running & Writing Tests
+=========================
```
```suggestion
=======================
Running & Writing Tests
=======================
```

CAM-Gerlach commented on tooling/clang.rst:
```diff
@@ -1,6 +1,8 @@
-***************************
+.. _clang:
+
+=============================
+ Dynamic Analysis with Clang
+=============================
```
```suggestion
===========================
Dynamic Analysis with Clang
===========================
```

CAM-Gerlach commented on tooling/index.rst:
```diff
@@ -0,0 +1,10 @@
+==================
+ Advanced Tooling
+==================
```
```suggestion
==============
Advanced Tools
==============
```

Shorter, simpler and more appropriate, since tooling implies a collection or system of tools/infrastructure, whereas these are individual, specific, niche tools.

CAM-Gerlach commented on triaging/gh-faq.rst:
```diff
@@ -1,11 +1,13 @@
+.. _gh-faq:
+
 GitHub issues for BPO users
 ===========================
```
```suggestion
GitHub Issues for BPO Users
===========================
```

Title case

CAM-Gerlach commented on triaging/index.rst:
```diff
@@ -0,0 +1,13 @@
+===========================
+ Issue Tracking and Triage
+===========================
```
```suggestion
===================
Issues and Triaging
===================
```

"Tracking" doesn't really add any informative value here, and triaging also includes PRs (which one of the two sections discusses, albeit with outdated terminology); otherwise, it would make more sense to call this subdirectory `issues` rather than triaging, as that would be the more general and appropriate term.

CAM-Gerlach commented on triaging/tracker.rst:
```diff
@@ -1,6 +1,6 @@
-==============
-Issue Tracking
-==============
+===============
+ Issue Tracker
+===============
```
.. _tracker:
.. _issue-tracker:
.. _issues:

```suggestion
=============
Issue Tracker
=============
```

Move this up here where it belongs, and add aliases used elsewhere for links

CAM-Gerlach commented on triaging/triageteam.rst:
```diff
@@ -0,0 +1,102 @@
+.. _triagers:
+.. _triageteam:
+
+====================
+ Python Triage Team
+====================
```
```suggestion
==================
Python Triage Team
==================
```

CAM-Gerlach commented on triaging/tracker.rst:
```diff
@@ -1,6 +1,6 @@
+===============
+ Issue Tracker
+===============
 
 .. _tracker:
```
```suggestion
```

Moved

CAM-Gerlach commented on triaging/triageteam.rst:
```diff
@@ -0,0 +1,102 @@
+requests. This helps contributors receive timely feedback and enables core
+developers to focus on reviewed items which reduces their workload. The
+expectations of this role expand upon the "Triager" role on the
+`issue tracker`_. The responsibilities listed below are primarily centered
```
```suggestion
:ref:`issue-tracker`. The responsibilities listed below are primarily centered
```

CAM-Gerlach commented on triaging/triageteam.rst:
```diff
@@ -0,0 +1,102 @@
+
+When you have consistently shown the ability to properly
+help triage issues without guidance, you may request that you
+be given the "Triager" role on the `issue tracker`_. You can make the request
```
```suggestion
be given the "Triager" role on the :ref:`issue-tracker`. You can make the request
```

CAM-Gerlach commented on triaging/triageteam.rst:
```diff
@@ -0,0 +1,102 @@
+For every new triager, it would be great to announce them in the python-committers
+mailing list and core-workflow category in Discourse. `Example announcement
+<https://discuss.python.org/t/abhilash-raj-has-been-granted-triage-role-on-github/2089>`_.
+
+.. _issue tracker: https://devguide.python.org/tracker/
```
```suggestion
```

Broken now; switched to ref

CAM-Gerlach commented on triaging/triaging.rst:
```diff
@@ -0,0 +1,107 @@
+ Triaging an Issue
+===================
+
+This section of the devguide documents the `issue tracker`_ for users
```
```suggestion
This section of the devguide documents the :ref:`issue-tracker` for users
```

CAM-Gerlach commented on triaging/triaging.rst:
```diff
@@ -0,0 +1,107 @@
+This section of the devguide documents the `issue tracker`_ for users
+and developers.
+
+Contributors with the Triager role on the `issue tracker`_ can triage issues
```
```suggestion
Contributors with the Triager role on the issue tracker can triage issues
```

Linked in previous sentence.

CAM-Gerlach commented on triaging/triaging.rst:
```diff
@@ -0,0 +1,107 @@
+    - Keywords
+* (Optional) Leave a brief comment about the proposed next action needed. If
+  there is a long message list, a summary can be very helpful.
+
+.. _issue tracker: https://devguide.python.org/tracker/
```
```suggestion
```

Broken; use an actual ref to fix

CAM-Gerlach commented on triaging/triaging.rst:
```diff
@@ -0,0 +1,107 @@
+contribute is to help triage issues. Do realize, though, that experience
+working on Python is needed in order to effectively help triage.
+
+Around the clock, new issues are being opened on the `issue tracker`_ and
```
```suggestion
Around the clock, new issues are being opened on the :ref:`issue-tracker` and
```

CAM-Gerlach reviewed (changes requested):
High-level feedback:

* [ ] The topic index appendix is now wholly redundant, since the ToC/sidebar categories effectively duplicate it (besides the docs that the former is missing), so it should just be removed—its one more place to maintain and possibly get out of sync, and doesn't serve any purpose.
* [ ] As discussed, the Experts Index should either be moved back to the Core Developers section (between the developer log and the motivations and affiliation), or if justifiably broad enough, made its own top-level section (i.e. where it was before) at the end of the devguide (where a top-level index belongs)
* [ ] I couldn't make individual comments since they were moved with no changes, but some of the titles should also be tweaked:
    * [ ] `Getting Started` -> `Setup and Building` (Much more specific description of its main contents, and frees up the name for the category title)
    * [ ] `Updating standard library extension modules` -> `Stdlib Extension Modules` (Better reflect the actual content as its as much about writing as "updating" them, make Title Case and abbreviate stdlib to make shorter and match the other title)
    * [ ] `Porting Python to a new platform` -> `Porting to a New Platform` (make Title Case and remove redundant and lengthy "Python")
    * [ ] `Documentation Markup` -> `reStructuredText Markup` ("Documentation" is already the section title and this makes more clear that this is a general reST guide rather than custom constructs added to the Python docs)
    * [ ] `Running a buildbot worker` -> `Running a 
    * [ ] `Exploring CPython's Internals` -> `Exploring the Internals`; `Guide to CPython's Parser` -> `Guide to the Parser`; `Design of CPython's Compiler` -> `Design of the Compiler` (shorter and avoid repeating CPython between each other and the category title)
    * [ ] `Core Developer Motivations and Affiliations` -> `Motivations and Affiliations` (since restating "Core Developers" sounds redundant and takes up a lot of needless space
    * [ ] `gdb Support` -> `GDB Support` (Use correct capitalization from [GDB website](https://www.sourceware.org/gdb/) and avoid looking strange)
* [ ] If we're going to be moving everything anyway, switching to `ref` for perma-links and retaining the old names as redirects, we may as well choose clearer, more appropriate and more descriptive file names while we have the chance. I was going to list them here, but I decided to wait until the other changes go in to avoid breaking existing suggestions, and also to be honest its 3 am and I'm exhausted from reviewing this for the past many hours, heh

AA-Turner commented on index.rst:
```diff
@@ -297,40 +297,14 @@ Full Table of Contents
-   buildworker
-   motivations
-   gitbootcamp
+   starting/index
+   developing/index
+   triaging/index
+   documenting/index
+   testing/index
+   committing/index
+   exploring/index
+   tooling/index
```
> I admire the parallelism between these (though `tooling` is a false friend to the others)

I tried... tooling was the hardest and I wasn't that happy with it. I'll do the rename but to preserve a good history I'll rebase and force push with it after all your other comments.

A

CAM-Gerlach replied:
Yeah, I felt really bad since I knew you must have put a good deal of thought into this, and I've tried for the exact same sort of thing in the past and been really disappointed if I couldn't get it to work out.

AA-Turner commented on starting/index.rst:
```diff
@@ -0,0 +1,12 @@
+.. toctree::
+   :maxdepth: 1
+
+   setup
+   help
+   pullrequest
+   fixingissues
+   gitbootcamp
```
Agree on switching issue/PR -- I originally put git last as it seemed more of an appendix topic, but for a entirely green contributor I agree you'd want it before you create a PR.

A

AA-Turner commented on testing/buildworker.rst:
```diff
@@ -1,8 +1,8 @@
 
-Running a buildbot worker
-=========================
+===========================
+ Running a buildbot worker
+===========================
```
I used "New Buildbot Workers".

A

AA-Turner commented:
> If we're going to be moving everything ... we may as well choose [better] file names

Implemented all your other changes (I went with a slightly different title for a few files). What filenames would you suggest?

Currently there is a bent towards `joinedupwords`, but `garbage_collector`, `c-api`, and `gh-(faq|labels)` are exceptions. I'd suggest preferring some form of spacing character for filenames of more than two words, and potentially for filenames of two words.

We should also update the directory names, as my `[a-z]ing` idea didn't work out. This should be possible without too much pain with `git filter-repo`.

A

hugovk commented:
> Currently there is a bent towards `joinedupwords`, but `garbage_collector`, `c-api`, and `gh-(faq|labels)` are exceptions. I'd suggest preferring some form of spacing character for filenames of more than two words, and potentially for filenames of two words.

Let's use a hyphen, because you can use alt+arrow key (on Mac, ctrl+arrow on Windows?) to jump between words. Underscore is treated as a letter.

CAM-Gerlach commented on exploring/index.rst:
```diff
@@ -0,0 +1,11 @@
+================================
+ Exploring Internals and Design
+================================
```
Unfortunately, despite this being my own review suggestion, I can't mark this as resolved without write on the repo, weirdly (but you can).

CAM-Gerlach commented:
> Implemented all your other changes (I went with a slightly different title for a few files).

Thanks so much, much appreciated! :tada: 

> What filenames would you suggest?

This is what I would suggest, following the discussion above (you could add ref aliases for the new names, but leave the existing ones obv). The main intent is to more helpfully describe the document's contents, clearly disambiguate it from others and ensure it is a good match for the document title, to aid navigating to/linking the correct one in the source.

<details>

<summary>Filename suggestions here</summary>

Core dev:
* `committing` -> `accepting-prs`
* `coredev` -> `becoming-coredev`
* `developers` -> `developer-log`

Development workflow:
* `communication` -> `communication-channels`
* `devcycle` -> `dev-cycle`
* `extensions` -> `extension-modules`
* `langchanges` -> `lang-changes`
* `stdlibchanges` -> `stdlib` (to parallel `extensions`)

CPython's internels:
* `garbage_collector` -> `garbage-collector`

 Getting Started:
* `fixingissues` -> `fixing-issues`
* `gitbootcamp` -> `git-bootcamp`
* `help` -> `getting-help`
* `pullrequest` -> `pull-request-lifecycle`
* `setup` -> `setup-building`

Testing:
* `buildworker` -> `setup-new-buildbot`
* `runtests` -> `run-write-tests`
* `silencewarnings` -> `silence-warnings`

Triaging:
* `tracker` -> `issue-tracker`
* `triageteam` -> `triage-team`

</details>

> Currently there is a bent towards joinedupwords, but garbage_collector, c-api, and gh-(faq|labels) are exceptions. I'd suggest preferring some form of spacing character for filenames of more than two words, and potentially for filenames of two words.

Definitely agreed (was going to go with `-`), that's a significant part of what I was thinking as it greatly improves readability.

> Let's use a hyphen, because you can use alt+arrow key (on Mac, ctrl+arrow on Windows?) to jump between words. Underscore is treated as a letter.

My plans as well :+1:

AA-Turner commented:
Clearly I messed something up when pushing after doing the filter-repo steps, I thought I'd rebased on upstream properly to retain the history. GitHub won't let me re-open the PR, so opening a new one.

A
